### ISSUE TITLE
[Improvement]: Add mysql dependency by maven profile

### ISSUE THREAD

**Issue by zhangwl9** [apache/amoro#3352]
### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/amoro/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

Currently, Amoro can choose  MySQL or Derby as the metadata database. However, when users choose MySQL, they need to manually obtain the dependency mysql-connector-j-XX.jar and download it to save in the "lib" directory. This behavior is rather cumbersome.To solve the problem, we would like to add the mysql-connector-j-XX.jar dependency in Maven with a default 'provided' scope, and then change it to 'compile' by activating a profile to include the dependency.

### How should we improve?

After this solution takes effect, we can build and enable mysql-connector-j: mvn clean package -DskipTests -Pmysql-j
the current version of mysql-connector-j used against MySQL Server version 8.0 and later

By default, the MySQL dependencies are not added to Amoro. When users choose to use MySQL as the metadata database, they can opt to add the MySQL dependencies to Amoro's lib directory in the above ways.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by zhangwl9** [comment]
@klion26 @czy006 Could you please take a look at this when you're free, thanks

**Comment by justinmclean** [comment]
MySQL's license is not compatible with the Apache license, so you should not be this. Can you please expand on "Currently Amoro frequently uses MySQL as a metadata database," because if it is used by most people, then is it not an optional dependency. See https://www.apache.org/legal/resolved.html#optional

**Comment by klion26** [comment]
@justinmclean Thanks for reminding us about the license information. Currently, Amoro stores the information used for ha/upgrade in a database(which can be MySql or Postgresql for now), and we do not include MySQL when releasing.

This issue here wants to add a profile that can be activated(default value is off) by the user who uses Mysql as the database to store their data used by Amoro, the dependency of MySQL would not been included in the Apache Amoro distribution. In other words, this issue did not change the content for the Apache Amoro distribution, it only added a feature for the users to easily package the MySQL dependency into their inner distribution or in their development env.

Does this align with the Apache Linces? or we can just add a `provided` dependency of MySql in our pom.xml, and can' support such profile to change the dependency scope from `provided` to `compile` in the pom.xml(with default value to `provided` and releasing with `provided` scope)

cc @zhoujinsong please correct me if there is something wrong

**Comment by justinmclean** [comment]
The issue is that it seems like it is not an optional dependency, so no it is not in line with ASF policy, as the majority of users would use it. ASF products can not depend on software with a GPL license [1], so it is an issue just depending on it and not including it. Even if that was not the case, the user needs a lot more warning that what they are going to do is incompatible with the Apache license.

1. https://www.apache.org/licenses/GPL-compatibility.html

**Comment by klion26** [comment]
@justinmclean Currently, there are some docs [1] for the users who use MySQL as the database(show that they need to download the MySQL dependency.  (users can choose the database from Derby/PostgreSQL/MySQL)

> You can use MySQL/PostgreSQL as the system database instead of the default Derby.
> If you would like to use MySQL as the system database, you need to manually download the [MySQL JDBC Connector](https://repo1.maven.org/maven2/com/mysql/mysql-connector-j/8.1.0/mysql-connector-j-8.1.0.jar) and move it into the {AMORO_HOME}/lib/ directory. You can use the following command to complete these operations:

I checked the Flink JDBC connector before reviewing this change, it is released without the MySQL dependency, and the doc[2] says that users need to download the dependency manually.

> The JDBC connector is not part of the binary distribution. See how to link with it for cluster execution [here](https://nightlies.apache.org/flink/flink-docs-release-1.20/docs/dev/configuration/overview/).
> A driver dependency is also required to connect to a specified database. Here are drivers currently supported:

Are there any best practices for Amoro use cases for MySQL dependency here, thanks.  

[1] https://amoro.apache.org/docs/latest/deployment/#configure-system-database
[2] https://nightlies.apache.org/flink/flink-docs-release-1.20/docs/connectors/table/jdbc/#dependencies

**Comment by justinmclean** [comment]
I feel you are misisng the main point of this conversation. What was stated above implies this dependency is not truly optional; is this the case or not? If this is not the case, how are users informed that MySQL is incompatible with the Apache license?

**Comment by klion26** [comment]
@justinmclean The MySQL dependency for Amoro is optional. Amoro would not be dependent on any MySQL-specific function; it only depends on common database connections and SQL grammar. Users can choose any database they are familiar with, MySQL is just one of the choices.

We'll use this issue to add some more information about the MySQL dependency license to the users
- move the whole dependency of MySQL to the optional profile
- add license information about MySQL in the **Amoro docs/pom profile/[build guild](https://github.com/apache/amoro?tab=readme-ov-file#building)** to inform users that MySQL license is not compatible with the Apache License

what do you think about this, thanks.

**Comment by justinmclean** [comment]
Just because a user can choose doesn't mean it is optional in teh ASF sense. Please answer the questions asked above.

**Comment by zhoujinsong** [comment]
Thanks for pushing forward this issue. @klion26 And thanks for providing the license information. @justinmclean 

I think there are some erros with the description of this issue like "Currently Amoro frequently uses MySQL as a metadata database". We have **NEVER** collected information about what database Amoro users used and come to the conclusion that most users use MySQL as Amoro's meta store.

Users can use any database as Amoro's meta store as it provides a JDBC implementation and Amoro will not rely on the unique features of any specific database.

We may give some guide for users who used MySQL as its meta store and profile is fine to me as if we should tell them the license incompatible issues in the profile description.

**Comment by justinmclean** [comment]
If most users rely on MySQL as Amoro's meta store, then it can't be considered an optional dependency. However, if MySQL usage is uncommon, there wouldn't be a need for this issue or PR. This creates a contradiction: either MySQL is widely used or it isn’t. Which is the case?

Making installation easier for users also places greater responsibility on the project to ensure they fully understand the implications of their choice. In my opinion, simply documenting the license is insufficient. For example, you could display a clear warning during installation, explicitly stating that it is not compatible. Users should then confirm their understanding by responding Y/N to a prompt before proceeding with the installation.

**Comment by zhoujinsong** [comment]
Yes, having the users only able to install the MySQL drivers manually is what exactly our choice for now.

The difference is that before this PR, users needed to manually download this dependency and place it in the project, while after this PR, users needed to manually enable the profile and rebuild the project.

Do you think this change is okay? @justinmclean

**Comment by justinmclean** [comment]
No, I do not, as I explained above.

**Comment by zhoujinsong** [comment]
Sure, maybe we should roll back #3380 and keep the original method of manually downloading the MySQL driver.

HDYT? @klion26 @zhangwl9

**Comment by klion26** [comment]
@justinmclean @zhoujinsong Thanks for the detailed reply. I'm ok with the rollback.

**Comment by justinmclean** [comment]
Why would you point the users to software incompatible with the Apache license and put the project's license and compliance with ASF policy at risk? Sure, please document it, but it should not be prominent as otherwise, is not an optional dependency. Where it is documented, it must be made very clear that it's incompatible and the user uses at their own legal risk.

**Comment by XBaith** [comment]
> Sure, maybe we should roll back [#3380](https://github.com/apache/amoro/pull/3380) and keep the original method of manually downloading the MySQL driver.

+1 for revert the PR ASAP

**Comment by justinmclean** [comment]
Looking at the issue a bit further, it seems it deeper than this one issue. I can see several places where it is stated that users should use MySQL in production.
